src/v/iceberg/conversion/values_json.cc

```diff
@@ -52,13 +52,16 @@ convert_primitive(serde::json::parser& p, const primitive_type& ft) {
         }
         return iceberg::boolean_value(false);
     case token::value_int:
-        if (!std::holds_alternative<long_type>(ft)) {
+        if (std::holds_alternative<long_type>(ft)) {
+            return iceberg::long_value(p.value_int());
+        } else if (std::holds_alternative<double_type>(ft)) {
+            return iceberg::double_value(static_cast<double>(p.value_int()));
+        } else {
             throw value_conversion_exception(
               fmt::format(
                 "Mismatch json between json integer value and schema type: {}",
                 ft));
         }
-        return iceberg::long_value(p.value_int());
     case token::value_double:
```

Comment:
If we get `42.0` that should be convertable to schema int type

Reply:
https://json-schema.org/understanding-json-schema/reference/numeric#integer

TBH I think we can just accept any lossy conversion and just round to the nearest int. We're not doing validation here

Reply:
PTAL. I think this is reasonable. Rounding would be surprising.

I think we could do better in a couple of cases (i.e. parsing `9223372036854775807.0e0` (2^63-1) as integer) but seeing that other parsers (i.e rapidjson) don't care much I'm not too concerned for now. Java does it right btw.

Will probably return to it later with rework for the precision issues too.